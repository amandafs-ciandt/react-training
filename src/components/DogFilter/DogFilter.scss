.dog-filter {
  padding: 1rem;
  padding-bottom: 0;

  &__item {
    display: flex;
    align-items: center;
  }

  .MuiChip-root {
    height: 18px;
    width: 18px;
    margin-left: 0.25rem;
  }

  .MuiChip-label {
    padding: 2px 0;
    font-size: 11px;
  }
}
