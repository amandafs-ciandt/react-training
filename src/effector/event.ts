import { createEvent } from 'effector';
import { Dog, BeerForm } from '../shared/types';

export const setLoaded = createEvent<Boolean>('set loaded list');

export const loadList = createEvent<Dog[]>('load dogs list');
export const updateList = createEvent<Dog>('update dogs list');

export const selectDog = createEvent<Dog>('select dog');
export const scoldDog = createEvent<void>('scold dog');

export const toggleFilter = createEvent<string>('toggle filter');

export const submitForm = createEvent<BeerForm>('submit form');
