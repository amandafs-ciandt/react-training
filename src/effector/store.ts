import { createStore, combine } from 'effector';
import {
  loadList,
  setLoaded,
  selectDog,
  scoldDog,
  updateList,
  toggleFilter,
  submitForm,
  setField,
} from './event';
import { Dog, BreedFilter, BeerForm, FormEvent } from '../shared/types';

export const dogBreedFilter = createStore([
  { letter: 'a', selected: false },
  { letter: 'b', selected: false },
  { letter: 'c', selected: false },
  { letter: 'd', selected: false },
  { letter: 'e', selected: false },
  { letter: 'f', selected: false },
  { letter: 'g', selected: false },
  { letter: 'h', selected: false },
  { letter: 'i', selected: false },
  { letter: 'j', selected: false },
  { letter: 'k', selected: false },
  { letter: 'l', selected: false },
  { letter: 'm', selected: false },
  { letter: 'n', selected: false },
  { letter: 'o', selected: false },
  { letter: 'p', selected: false },
  { letter: 'q', selected: false },
  { letter: 'r', selected: false },
  { letter: 's', selected: false },
  { letter: 't', selected: false },
  { letter: 'u', selected: false },
  { letter: 'v', selected: false },
  { letter: 'w', selected: false },
  { letter: 'x', selected: false },
  { letter: 'y', selected: false },
  { letter: 'z', selected: false },
  { letter: '', selected: false },
]).on(toggleFilter, (list: BreedFilter[], selectedLetter: string) =>
  list.map((breedFilterItem: BreedFilter) => {
    return breedFilterItem.letter === selectedLetter
      ? {
          ...breedFilterItem,
          selected: true,
        }
      : {
          ...breedFilterItem,
          selected: false,
        };
  })
);

const filter = createStore('').on(
  dogBreedFilter,
  (_, list) =>
    list.find((breedFilter: BreedFilter) => breedFilter.selected)?.letter
);

const filterFn = filter.map((filter) => {
  if (filter === '') {
    return (dog: Dog) => true;
  } else {
    return (dog: Dog) => dog.breed.startsWith(filter);
  }
});

export const loaded = createStore<Boolean>(false).on(
  setLoaded,
  (_, isLoaded: Boolean) => isLoaded
);

export const selectedDog = createStore<Dog>({
  breed: '',
  image: '',
  scolded: 0,
})
  .on(selectDog, (_, newSelectedDog: Dog) => newSelectedDog)
  .on(scoldDog, (currentSelectedDog: Dog) => {
    const updatedDog = {
      ...currentSelectedDog,
      scolded: currentSelectedDog.scolded + 1,
    };

    updateList(updatedDog);

    return updatedDog;
  });

export const dogs = createStore<Dog[]>([])
  .on(loadList, (_, list: Dog[]) => list)
  .on(updateList, (list: Dog[], updatedDog: Dog) =>
    list.map((dog: Dog) => {
      if (dog.breed === updatedDog.breed) {
        return updatedDog;
      }

      return dog;
    })
  );

export const filtered = combine(dogs, filterFn, (list: Dog[], fn: any) =>
  list.filter(fn)
);

export const beerForm = createStore<BeerForm>({
  beerName: '',
  beerType: '',
  hasCorn: false,
  ingredients: '',
}).on(setField, (form: BeerForm, event: FormEvent) => {
  return { ...form, [event.name]: event.value };
});

export const validForm = beerForm.map(
  (form: BeerForm) =>
    form.beerName.length > 0 &&
    form.beerType.length > 0 &&
    form.ingredients.length > 0
);

export const beerFormik = createStore<BeerForm>({
  beerName: '',
  beerType: '',
  hasCorn: false,
  ingredients: '',
}).on(submitForm, (_, submittedData: any) => submittedData);
