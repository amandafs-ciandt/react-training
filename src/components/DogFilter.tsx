import React from 'react';
import _ from 'lodash';

import { RadioChangeEvent } from 'antd/lib/radio';
import { Badge, Radio } from 'antd';
import './DogFilter.scss';

import { useList } from 'effector-react';
import { dogBreedFilter } from '../effector/store';
import { toggleFilter } from '../effector/event';

const DogFilter = () => {
  const filterOptions = useList(dogBreedFilter, ({ letter }, index) =>
    letter !== '' ? (
      <Radio.Button key={index} value={letter}>
        {_.upperFirst(letter)}
        <Badge showZero count={0} style={{ marginLeft: '0.5rem' }} />
      </Radio.Button>
    ) : (
      <Radio.Button key={index} value={letter}>
        Clear
      </Radio.Button>
    )
  );

  const onChange = (e: RadioChangeEvent) => {
    e.preventDefault();
    toggleFilter(e.target.value);
  };

  return (
    <div className='dog-filter'>
      <Radio.Group onChange={onChange} size='middle'>
        {filterOptions}
      </Radio.Group>
    </div>
  );
};

export default DogFilter;
