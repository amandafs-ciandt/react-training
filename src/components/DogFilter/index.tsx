import React, { ChangeEvent } from 'react';
import _ from 'lodash';

import { RadioGroup, FormControlLabel, Radio, Chip } from '@material-ui/core';

import { useStore } from 'effector-react';
import { dogBreedFilter, dogs } from '../../effector/store';
import { toggleFilter } from '../../effector/event';

import { Dog } from '../../shared/types';

import './DogFilter.scss';

const DogFilter = () => {
  const dogList = useStore(dogs);
  const filterOptions = useStore(dogBreedFilter);

  const filterItem = (letter: string) => (
    <div className='dog-filter__item'>
      {_.upperFirst(letter)}{' '}
      <Chip
        label={
          dogList.filter((dog: Dog) => dog.breed.startsWith(letter)).length
        }
        data-testid='dog-item-scolded-value'
        color='secondary'
        size='small'
      />
    </div>
  );

  const breedFilter = filterOptions.map(({ letter }, index) => {
    if (letter !== '') {
      return (
        <FormControlLabel
          key={index}
          value={letter}
          control={<Radio />}
          label={filterItem(letter)}
          data-testid='dog-filter-option'
        />
      );
    } else {
      return (
        <FormControlLabel
          key={index}
          value={letter}
          control={<Radio />}
          label={<div className='dog-filter__item'>All</div>}
          data-testid='dog-filter-option'
        />
      );
    }
  });

  const onChange = (e: ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    toggleFilter(e.target.value);
  };

  return (
    <div className='dog-filter'>
      <RadioGroup row name='dog-filter' onChange={onChange}>
        {breedFilter}
      </RadioGroup>
    </div>
  );
};

export default DogFilter;
