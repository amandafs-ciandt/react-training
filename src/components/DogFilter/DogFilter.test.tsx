import React from 'react';
import { shallow, ShallowWrapper, mount, ReactWrapper } from 'enzyme';
import { when } from 'jest-when';

import * as effector from 'effector-react';
import { dogs, dogBreedFilter } from '../../effector/store';

import DogFilter from '.';

const dogList = [
  {
    breed: 'bulldog',
    image: 'https://images.dog.ceo/breeds/bulldog-french/n02108915_4474.jpg',
    scolded: 3,
  },
  {
    breed: 'beagle',
    image: 'https://images.dog.ceo/breeds/beagle/n02088364_12102.jpg',
    scolded: 1,
  },
  {
    breed: 'collie',
    image: 'https://images.dog.ceo/breeds/collie-border/n02106166_1460.jpg',
    scolded: 0,
  },
];

const filterOptions = [
  { letter: 'a', selected: false },
  { letter: 'b', selected: false },
  { letter: 'c', selected: false },
  { letter: '', selected: false },
];

const selectedDogSpy = jest.spyOn<any, any>(effector, 'useStore');
when(selectedDogSpy).calledWith(dogs).mockReturnValue(dogList);
when(selectedDogSpy).calledWith(dogBreedFilter).mockReturnValue(filterOptions);

const setUpShallowRendering = (): ShallowWrapper => {
  return shallow(<DogFilter />);
};

const setUpMountedRendering = (): ReactWrapper => {
  return mount(<DogFilter />);
};

describe('DogFilter', () => {
  let dogFilterComponent: ShallowWrapper | ReactWrapper;

  it('should render correctly', () => {
    dogFilterComponent = setUpShallowRendering();

    expect(dogFilterComponent).toMatchSnapshot();
  });

  it('should show all filter options correctly', () => {
    dogFilterComponent = setUpMountedRendering();

    const dogFilterOptions = dogFilterComponent.find('.dog-filter__item');

    expect(dogFilterOptions.length).toEqual(filterOptions.length);

    expect(dogFilterOptions.at(0).text()).toEqual('A 0');
    expect(dogFilterOptions.at(1).text()).toEqual('B 2');
    expect(dogFilterOptions.at(2).text()).toEqual('C 1');
    expect(dogFilterOptions.at(3).text()).toEqual('All');
  });

  it('should call toggle filter when selected an option', () => {
    //TODO
  });
});
