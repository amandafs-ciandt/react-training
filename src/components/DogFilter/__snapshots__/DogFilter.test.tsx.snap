// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogFilter should render correctly 1`] = `ShallowWrapper {}`;
